tests/Tokenizer/TokensAnalyzerTest.php

```diff
@@ -1909,6 +1909,44 @@ public static function provideIsBinaryOperator81Cases(): iterable
         ];
     }
 
+    /**
+     * @param array<int, bool> $expected
+     *
+     * @dataProvider provideIsBinaryOperator82Cases
+     *
+     * @requires PHP 8.2
+     */
+    public function testIsBinaryOperator82(array $expected, string $source): void
+    {
+        $tokensAnalyzer = new TokensAnalyzer(Tokens::fromCode($source));
+
+        foreach ($expected as $index => $isBinary) {
+            static::assertSame($isBinary, $tokensAnalyzer->isBinaryOperator($index));
+
+            if ($isBinary) {
+                static::assertFalse($tokensAnalyzer->isUnarySuccessorOperator($index));
+                static::assertFalse($tokensAnalyzer->isUnaryPredecessorOperator($index));
+            }
```

Comment:
Let's make use of this check, I've added quite tricky examples.

Reply:
Not sure about the changes you made. `foreach ($tokens as $index => $token)` has 2 flaws:

- `$token` is not used, so data provider could be simplified
- `foreach` iterates through all tokens, so `assertSame()` is called for each token which artificially inflates the number of assertions... Previous format was better in that regard, it was testing only operators that could represent either binary or type intersection/union.

The question is if we want to check that type unions/intersections are not determined as binary operator, or also we want to check that binary operator is found properly somewhere in between of DNFs 🙂.

Reply:
> `$token` is not used, so data provider could be simplified

I do not understand how data provider could be simplified because of that, can you explain?

> `foreach` iterates through all tokens, so `assertSame()` is called for each token which artificially inflates the number of assertions...

it also makes sure that all tokens are verified - so all tokens not listed in `$expected` are not binary operators.

Reply:
We can later refactor all these similar `testIsBinaryOperatorXX` tests, but let's focus on PHP 8.2 support here, not making tests perfect.